cli/util/upload_bin.go

```diff
@@ -197,129 +196,99 @@ func uploadBlocksAndIndexFiles(ctx *cli.Context, p *pool.Pool, rpc *rpcclient.Cl
+	for b := int(totalBatches) - 1; b >= 0; b-- {
+		start := uint(b) * batchSize
+		end := min(start+batchSize, currentBlockHeight)
+
+		// Collect blocks in [start, end).
+		blkErr := make(chan error)
+		items := searchObjects(ctx.Context, p, containerID, privKeys,
+			blockAttributeKey, start, end, maxParallelSearches, blkErr)
 
-		var lastItem *client.SearchResultItem
+		existing := make(map[uint]oid.ID)
 	loop:
 		for {
 			select {
-			case itm, ok := <-results:
+			case itm, ok := <-items:
 				if !ok {
 					break loop
 				}
-				lastItem = &itm
-			case err := <-errs:
+				idx, err := strconv.ParseUint(itm.Attributes[0], 10, 32)
 				if err != nil {
-					select {
-					case errCh <- fmt.Errorf("failed to search for index file: %w", err):
-					default:
-					}
+					return 0, nil, fmt.Errorf("failed to parse block index: %w", err)
 				}
-				break loop
+				existing[uint(idx)] = itm.ID
+			case err := <-blkErr:
+				return 0, nil, err
 			}
 		}
-		if lastItem != nil {
-			parsed, err := strconv.ParseUint(lastItem.Attributes[0], 10, 32)
-			if err != nil {
-				select {
-				case errCh <- fmt.Errorf("failed to parse index file ID: %w", err):
-				default:
-				}
-				return
-			}
-			existingIndex = parsed
-		}
-		fmt.Fprintf(ctx.App.Writer, "Current index files count: %d\n", existingIndex+1)
 
-		// Start block parsing goroutines.
-		var (
-			// processedIndices is a mapping from position in buffer to the block index.
-			// It prevents duplicates.
-			processedIndices sync.Map
-			wg               sync.WaitGroup
-			objCh            = make(chan client.SearchResultItem, 2*maxParallelSearches)
-		)
-		wg.Add(int(maxParallelSearches))
-		for range maxParallelSearches {
-			go func() {
-				defer wg.Done()
-				for obj := range objCh {
-					blockIndex, err := strconv.ParseUint(obj.Attributes[0], 10, 32)
-					if err != nil {
-						select {
-						case errCh <- fmt.Errorf("failed to get block index from object %s with attributes %s: %w", obj.ID, obj.Attributes, err):
-						default:
-						}
-						return
-					}
-					pos := uint(blockIndex) % indexFileSize
-					if _, ok := processedIndices.LoadOrStore(pos, blockIndex); !ok {
-						copy(buf[pos*oid.Size:], obj.ID[:])
-					}
-				}
-			}()
+		if uint(len(existing)) == end-start {
+			lastFullFound = true
+			lastFull = uint(b)
+			break
 		}
+	}
 
-		// Search for blocks within the index file range.
-		blkObjs := searchObjects(ctx.Context, p, containerID, privKeys, blockAttributeKey, uint(existingIndex)*indexFileSize, uint(existingIndex+1)*indexFileSize, maxParallelSearches, errCh)
-		for id := range blkObjs {
-			objCh <- id
-		}
-		close(objCh)
-		wg.Wait()
-	}()
+	var nextBatch uint
+	if lastFullFound {
+		nextBatch = lastFull + 1
+	} else {
+		nextBatch = 0
+	}
+	fmt.Fprintf(ctx.App.Writer, "Last fully uploaded batch: %d, next to upload: %d\n",
+		func() uint {
+			if lastFullFound {
+				return lastFull
+			}
+			return ^uint(0)
+		}(), nextBatch)
```

Comment:
Avoid such constructions, it complicates code for nothing. Use a separate variable, assign proper value to it then log.